segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -710,8 +717,11 @@ public void readTable(final WireCommands.ReadTable readTable) {
         val timer = new Timer();
         tableStore.get(segment, keys, TIMEOUT)
                 .thenAccept(values -> {
-                    connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment, getTableEntriesCommand(keys, values)));
-                    this.tableStatsRecorder.getKeys(readTable.getSegment(), keys.size(), timer.getElapsed());
+                    // NOTE: getTableEntriesCommand will truncate the result to prevent it from exceeding MAX_TABLE_RESPONSE_SIZE.
+                    // In such a situation, not all requested items will be returned to the caller; only the head of the list will.
+                    WireCommands.TableEntries response = getTableEntriesCommand(readTable.getRequestId(), keys, values);
```

Comment:
This seems particularly inefficient. Supposing there are N items to be retrieved and a fixed number fit in a batch this is O(n^2) because the first call gets all N and throws all but the first batch away. The second fetches N-batch and throws all but the next batch away etc.
We should figure out a way to pass down a limit to the get call, even if it is not exact. It's better to reply with more smaller messages which are efficient than maximizing our wirecommand size out in the reply.

Reply:
I agree with this. However I am not sure how easy it is to do this with the current `RawClient` that expects a single response for a request.

I filed #4936 to research a way to handle these situations. This may also help iterators too (i.e., to avoid a flurry of request-replies for each iterator iteration).

Reply:
I reworked this part. Check out the updated description for details.